### PR TITLE
Implemented FedBE

### PR DESCRIPTION
**Pull request by NisseOscar**

### CODE REVIEW

File: Algorithms/FedBe.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+from Servers.FedBeServer import FedBeServer
+from Clients.FedAvgClient import FedAvgClient
+from Algorithms.ABCAlgorithm import ABCAlgorithm
+
+class FedBe(ABCAlgorithm):
+    def __init__(self,dataloader,Model,loc_data,callbacks=None,n_clients=5,M=10,
```
Comment (MartinHex) [inline Algorithms/FedBe.py]: Arguably we only need default values either here or in the server. Keeping the default values at the server seem reasonable and ot make anything more than a default FedBE algorithm you'd access  the server specifically through the algorithm object.

File: Servers/FedBeServer.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
+from Servers.ABCServer import ABCServer
+import torch
+from torch.distributions import Normal
+from torch.nn import NLLLoss
+from torchcontrib.optim import SWA
+from torch.optim import SGD
+import random
+import numpy as np
+import warnings
+warnings.filterwarnings("ignore")
```
Comment (MartinHex) [inline Servers/FedBeServer.py]: 😄

File: Servers/FedBeServer.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            mu_n2 = (n*mu_n+w_flat)/(n+1)
+            M_n= M_n+(w_flat-mu_n)*(w_flat-mu_n2)
+            if(n!=0):
+                s_n= torch.sqrt(M_n/n)
+            mu_n = mu_n2
+            S.append(w)
+
+        # Add small variance to avoid zero variance
+        s_n = s_n+0.0000001
+        self.distribution = Normal(mu_n,s_n)
```
Comment (MartinHex) [inline Servers/FedBeServer.py:54]: Standardize mu towards the amount of data.

File: Servers/FedBeServer.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+                    # Add samples to map loss landscape only during first iteration
+                    if(len(X)!=len(p)):
+                        X.append(x)
+            res = np.array(res)
+            if(p.size == 0):
+                p = np.zeros(len(res))
+            p+=res
+
+        p = p/len(S)
+        T = [(x_j,p_j) for x_j,p_j in zip(X,p)]
```
Comment (MartinHex) [inline Servers/FedBeServer.py]: Change name of T to not overlap with True

File: Servers/FedBeServer.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        # initiate and run SWA
+        self.model.set_weights(mu_r)
+        self.model.train()
+        base_opt = torch.optim.SGD(self.model.parameters(), lr=self.swa_lr1)
+        opt = SWA(base_opt,  swa_freq=self.swa_freq, swa_lr=self.swa_lr2)
+        for i in range(self.swa_epochs):
+             for batch in T_batched:
+                 opt.zero_grad()
+                 X_batched = [b[0] for b in batch]
+                 p_batched = torch.tensor([b[1] for b in batch])
```
Comment (MartinHex) [inline Servers/FedBeServer.py]: Change p to tensor directly, and dont import numpy.

File: Servers/FedBeServer.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+                 pred = self.model.forward(torch.stack(X_batched))[0]
+                 loss=torch.mean(pred)*0-torch.mean(p_batched)
```
Comment (MartinHex) [inline Servers/FedBeServer.py:117]: Comment on this.

File: Servers/FedBeServer.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+                 loss.backward()
+                 opt.step()
+             # if self.verbose: print('FedBE: Epoch %i: loss: %.4f'%(i,-loss.item()))
```
Comment (MartinHex) [inline Servers/FedBeServer.py]: Should this be commeted?

**Review by MartinHex — CHANGES_REQUESTED** [review]

File: Servers/FedBeServer.py (hunk 3 above)
Comment (NisseOscar) [reply Servers/FedBeServer.py:54]: This is not possible considering current design. Possible add a client.data_size if we want to add this feature.

File: Servers/FedBeServer.py (hunk 2 above)
Comment (NisseOscar) [reply Servers/FedBeServer.py]: Removed in latest commit fixup.

File: Algorithms/FedBe.py (hunk 1 above)
Comment (NisseOscar) [reply Algorithms/FedBe.py]: Removed in latest commit.

File: Servers/FedBeServer.py (hunk 4 above)
Comment (NisseOscar) [reply Servers/FedBeServer.py]: Changed to T_tau in latest commit.

File: Servers/FedBeServer.py (hunk 5 above)
Comment (NisseOscar) [reply Servers/FedBeServer.py]: p is now a tensor.

**Review by MartinHex — APPROVED** [review]
Looks good! 👀

**Pull request merged**